internal/config/ociartifact/ociartifact.go

```diff
@@ -24,54 +29,90 @@ func New() Impl {
 
 // Artifact can be used to manage OCI artifacts.
 type Artifact struct {
-	// MountPath is the local path containing the artifact data.
-	MountPath string
-
-	// Cleanup has to be called if the artifact is not used any more.
-	Cleanup func()
+	// Data is the actual artifact content.
+	Data []byte
 }
 
 // defaultImpl is the default implementation for the OCI artifact handling.
 type defaultImpl struct{}
 
-// Pull downloads and mounts the artifact content by using the provided ref.
-func (*defaultImpl) Pull(ctx context.Context, sys *types.SystemContext, ref string) (*Artifact, error) {
-	log.Infof(ctx, "Pulling OCI artifact from ref: %s", ref)
+const (
+	maxArtifactSize = 1024 * 1024 // 1 MiB
+)
+
+// Pull downloads and mounts the artifact content by using the provided image name.
+func (*defaultImpl) Pull(
+	ctx context.Context,
+	sys *types.SystemContext,
+	img string,
+	enforceConfigMediaType string,
+) (*Artifact, error) {
+	log.Infof(ctx, "Pulling OCI artifact from ref: %s", img)
 
-	storeOpts, err := storage.DefaultStoreOptions(false, 0)
+	name, err := libimage.NormalizeName(img)
 	if err != nil {
-		return nil, fmt.Errorf("get default storage options: %w", err)
+		return nil, fmt.Errorf("parse image name: %w", err)
 	}
 
-	store, err := storage.GetStore(storeOpts)
+	ref, err := docker.NewReference(name)
 	if err != nil {
-		return nil, fmt.Errorf("get container storage: %w", err)
+		return nil, fmt.Errorf("create docker reference: %w", err)
 	}
 
-	runtime, err := libimage.RuntimeFromStore(store, &libimage.RuntimeOptions{SystemContext: sys})
+	src, err := ref.NewImageSource(ctx, sys)
 	if err != nil {
-		return nil, fmt.Errorf("create libimage runtime: %w", err)
+		return nil, fmt.Errorf("build image source: %w", err)
 	}
 
-	images, err := runtime.Pull(ctx, ref, config.PullPolicyAlways, &libimage.PullOptions{})
+	manifestBytes, mimeType, err := src.GetManifest(ctx, nil)
 	if err != nil {
-		return nil, fmt.Errorf("pull OCI artifact: %w", err)
+		return nil, fmt.Errorf("get manifest: %w", err)
+	}
+
+	parsedManifest, err := manifest.FromBlob(manifestBytes, mimeType)
+	if err != nil {
+		return nil, fmt.Errorf("parse manifest: %w", err)
+	}
+	if enforceConfigMediaType != "" && parsedManifest.ConfigInfo().MediaType != enforceConfigMediaType {
+		return nil, fmt.Errorf(
+			"wrong config media type %q, requires %q",
+			parsedManifest.ConfigInfo().MediaType, enforceConfigMediaType,
+		)
 	}
-	image := images[0]
 
-	mountPath, err := image.Mount(ctx, nil, "")
+	layers := parsedManifest.LayerInfos()
+	if len(layers) < 1 {
+		return nil, errors.New("artifact needs at least one layer")
+	}
+	layer := layers[0]
+
+	bic := blobinfocache.DefaultCache(sys)
+	rc, _, err := src.GetBlob(ctx, layer.BlobInfo, bic)
+	if err != nil {
+		return nil, fmt.Errorf("get layer blob: %w", err)
+	}
+	defer rc.Close()
+
+	layerBytes, err := io.ReadAll(io.LimitReader(rc, maxArtifactSize))
+	if len(layerBytes) == maxArtifactSize {
```

Comment:
When my image happens to be exactly 1 MiB, it will be a failure?

The `GetBlob()` function returns size if `Content-Size` header is present in the response (we assume remote server is well-behaved), so we could leverage this first to make a check and fail early (no need to actually read anything then), and if we got `-1`, then we could rely on this check here.

It would be nice to make this check a > 1 MiB, if at all possible.

Thoughts?

Reply:
(Compare the implementation in `c/image/internal/iolimits`.)

Reply:
Done, fixed.